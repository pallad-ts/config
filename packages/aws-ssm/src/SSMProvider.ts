import { left, right } from "@sweet-monads/either";
// eslint-disable-next-line @typescript-eslint/naming-convention
import DataLoader from "dataloader";

import { AsyncProvider, Provider, ValueNotAvailable } from "@pallad/config";

export class SSMProvider extends AsyncProvider<SSMProvider.Value> {
    constructor(
        private key: string,
        private dataLoader: DataLoader<string, SSMProvider.Value | undefined>
    ) {
        super();
    }

    resolveValueAsync(): Promise<Provider.Value<SSMProvider.Value>> {
        return this.dataLoader
            .load(this.key)
            .then(value => {
                if (value === undefined) {
                    return left<Provider.Fail, SSMProvider.Value>(new ValueNotAvailable(`SSM: ${this.key}`));
                }
                return right<Provider.Fail, SSMProvider.Value>(value);
            })
            .catch(x => {
                return left<Provider.Fail, SSMProvider.Value>(x);
            });
    }
}

export namespace SSMProvider {
    export type Value = string | string[];
}
