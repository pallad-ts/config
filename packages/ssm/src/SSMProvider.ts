import {Provider} from '@pallad/config';
import {DataLoader} from './index';

export class SSMProvider extends Provider<string | string[]> {

    private cache?: Promise<SSMProvider.Value | undefined>;

    constructor(private key: string,
                private dataLoader: DataLoader<string, SSMProvider.Value | undefined>) {
        super()
    }

    getDescription(): string {
        return `SSM (${this.key})`;
    }

    async isAvailable(): Promise<boolean> {
        return (await this.loadValue()) !== undefined;
    }

    private loadValue() {
        if (!this.cache) {
            this.cache = this.dataLoader.load(this.key);
        }
        return this.cache;
    }

    protected retrieveValue(): Promise<any> {
        return this.loadValue();
    }
}

export namespace SSMProvider {
    export type Value = string | string[];
}
