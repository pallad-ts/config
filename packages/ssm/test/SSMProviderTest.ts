import * as sinon from 'sinon';
import DataLoader = require("dataloader");
import {SSMProvider} from "@src/SSMProvider";

describe('SSMProviderTest', () => {
    let dataLoader: sinon.SinonStubbedInstance<DataLoader<string, SSMProvider.Value | undefined>>;

    const KEY = 'foo';
    const RESULT = 'bar';

    function stubSSMResult(key: string, result: SSMProvider.Value) {
        dataLoader.load
            .withArgs(key)
            .resolves(result);
    }

    beforeEach(() => {
        dataLoader = sinon.createStubInstance(DataLoader);
    });

    describe('is available', () => {
        it('true if given ssm key exists', () => {
            const dep = new SSMProvider(KEY, dataLoader);
            stubSSMResult(KEY, RESULT);

            return expect(dep.isAvailable())
                .resolves
                .toEqual(true);
        });

        it('false if given ssm key does not exist', () => {
            const dep = new SSMProvider(KEY, dataLoader);
            return expect(dep.isAvailable())
                .resolves
                .toEqual(false);
        });
    });

    it('getting description', () => {
        const dep = new SSMProvider(KEY, dataLoader);
        expect(dep.getDescription())
            .toMatchSnapshot();
    });

    describe('getting value', () => {
        it('success', () => {
            const dep = new SSMProvider(KEY, dataLoader);
            stubSSMResult(KEY, RESULT);
            return expect(dep.getValue())
                .resolves
                .toEqual(RESULT);
        });

        it('failure', () => {
            const dep = new SSMProvider(KEY, dataLoader);

            return expect(dep.getValue())
                .rejects
                .toThrowErrorMatchingSnapshot()
        });
    });
});
