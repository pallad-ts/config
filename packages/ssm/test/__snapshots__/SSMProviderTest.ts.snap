// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSMProviderTest getting description 1`] = `"SSM (foo)"`;

exports[`SSMProviderTest getting value failure 1`] = `"Config value not available: SSM (foo)"`;
